nucleus/connection.py

```diff
@@ -0,0 +1,71 @@
+import requests
+import logging
+
+from .constants import DEFAULT_NETWORK_TIMEOUT_SEC
+from .errors import NucleusAPIError
+
+logger = logging.getLogger(__name__)
+logging.basicConfig()
+logging.getLogger(requests.packages.urllib3.__package__).setLevel(
+    logging.ERROR
+)
+
+
+class RetryStrategy:
+    statuses = {503, 504}
+    sleep_times = [1, 3, 9]
+
+
+class Connection:
+    """Wrapper of HTTP requests to the Nucleus endpoint."""
+
+    def __init__(self, api_key: str, endpoint: str = None):
+        self.api_key = api_key
+        self.endpoint = endpoint
+
+    def make_request(
+        self, payload: dict, route: str, requests_command=requests.post
+    ) -> dict:
+        """
+        Makes a request to Nucleus endpoint and logs a warning if not
+        successful.
+
+        :param payload: given payload
+        :param route: route for the request
+        :param requests_command: requests.post, requests.get, requests.delete
+        :return: response JSON
+        """
+        endpoint = f"{self.endpoint}/{route}"
+
+        logger.info("Make request to %s", endpoint)
+
+        for retry_wait_time in RetryStrategy.sleep_times:
```

Comment:
I would probably just import some retry library to help with this. I see that we use `tenacity` in `models` FWIW. Doing an exponential backoff with base 3 instead of base 2 is a bit quirky 😛

Reply:
I'll add a comment for this as a TODO. I don't want to lump that into this PR - this code was directly copied from the Nucleus base python client, so I think that we could introduce a change there in an individual PR.

Reply:
Created this ticket to track: https://app.shortcut.com/scaleai/story/281630/use-external-library-for-http-retry-logic